hail/src/main/scala/is/hail/expr/types/physical/PStruct.scala

```diff
@@ -18,21 +18,35 @@ trait PStruct extends PBaseStruct {
     CodeOrdering.rowOrdering(this, other.asInstanceOf[PStruct], mb, so)
   }
 
-  def deleteField(key: String): PStruct
+  final def deleteField(key: String): PCanonicalStruct = {
+    assert(fieldIdx.contains(key))
+    val index = fieldIdx(key)
+    val newFields = Array.fill[PField](fields.length - 1)(null)
+    for (i <- 0 until index)
+      newFields(i) = fields(i)
+    for (i <- index + 1 until fields.length)
+      newFields(i - 1) = fields(i).copy(index = i - 1)
+    PCanonicalStruct(newFields, required)
+  }
 
-  def appendKey(key: String, sig: PType): PStruct
+  final def appendKey(key: String, sig: PType): PCanonicalStruct = {
+    assert(!fieldIdx.contains(key))
+    val newFields = Array.fill[PField](fields.length + 1)(null)
+    for (i <- fields.indices)
+      newFields(i) = fields(i)
+    newFields(fields.length) = PField(key, sig, fields.length)
+    PCanonicalStruct(newFields, required)
+  }
 
   def rename(m: Map[String, String]): PStruct
 
-  def ++(that: PStruct): PStruct
-
   def identBase: String = "tuple"
 
-  def selectFields(names: Seq[String]): PStruct
+  final def selectFields(names: Seq[String]): PCanonicalStruct = PCanonicalStruct(required, names.map(f => f -> field(f).typ): _*)
 
-  def dropFields(names: Set[String]): PStruct
+  final def dropFields(names: Set[String]): PCanonicalStruct = selectFields(fieldNames.filter(!names.contains(_)))
 
-  def typeAfterSelect(keep: IndexedSeq[Int]): PStruct
+  final def typeAfterSelect(keep: IndexedSeq[Int]): PCanonicalStruct = PCanonicalStruct(required, keep.map(i => fieldNames(i) -> types(i)): _*)
```

Comment:
This would make for slightly strange interface I think.

```scala
val ps = PSubsetStruct(somePStruct, "a", "b");
val fewerFields = ps.selectFields("a")
```

`fewerFields` is now a PCanonicalStruct. I would expect it to be a PSubsetStruct with no field "a", and the same PStruct that was PSubsetStruct during construction.

The other alternative that makes sense to me is a PSubsetStruct with an underlying PStruct that also has no field "a". 

I think it will be confusing if the return is instead a PCanonicalStruct.

Reply:
this function should be removed, but all of its usages now are modifying the type then reconstruct the value with a RVB. Returning a PSubsetStruct would be incorrect, because that is not constructible. PSubsetStruct.typeAfterSelect should return a constructible PCanonicalStruct in the current system. 

Some of the usages can definitely use a PSubsetStruct and avoid copying values, though, once we support non-constructible types!

Reply:
I was speaking about all of the final methods, not just typeAfterSelect. Can we move these methods to PCanonicalStruct? It will be confusing to have methods like "dropFields" or "selectFields", whose only action is to modify which fields the type tracks (and can therefore could work on any PStruct, not just a canonical or constructible one), return a different type than the one it was called on.

Reply:
if we move them to PCanonicalStruct, we'll need to cast to use them. Those casts will be invalid if the type is, say, a PSubsetStruct.

Do you want me to rename these methods to `constructibleSelectFields` or something?

Reply:
Nah, that seems like unnecessary work. I'll approve as is, if we have problems we'll have this conversation in the back of our minds.

Reply:
👍 thanks!